Sources/PostgresNIO/New/Messages/DataRow.swift

```diff
@@ -1,34 +1,132 @@
-            
-            for _ in 0..<columnCount {
-                try buffer.ensureAtLeastNBytesRemaining(2)
-                let bufferLength = Int(buffer.readInteger(as: Int32.self)!)
-                
-                guard bufferLength >= 0 else {
-                    result.append(nil)
-                    continue
-                }
-                
-                try buffer.ensureAtLeastNBytesRemaining(bufferLength)
-                let columnBuffer = buffer.readSlice(length: Int(bufferLength))!
-                
-                result.append(columnBuffer)
+            guard bufferLength >= 0 else {
+                // if buffer length is negative, this means that the value is null
+                continue
             }
             
-            return DataRow(columns: result)
+            try buffer.ensureAtLeastNBytesRemaining(bufferLength)
+            buffer.moveReaderIndex(forwardBy: bufferLength)
+        }
+        
+        try buffer.ensureExactNBytesRemaining(0)
+        
+        buffer.moveReaderIndex(to: firstColumnIndex)
+        let columnSlice = buffer.readSlice(length: buffer.readableBytes)!
+        return DataRow(columnCount: columnCount, bytes: columnSlice)
+    }
+}
+
+extension DataRow: Sequence {
+    typealias Element = ByteBuffer?
+    
+    // There is no contiguous storage available... Sadly
+    func withContiguousStorageIfAvailable<R>(_ body: (UnsafeBufferPointer<ByteBuffer?>) throws -> R) rethrows -> R? {
+        nil
+    }
+}
+
+extension DataRow: Collection {
+    
+    struct ColumnIndex: Comparable {
+        var offset: Int
+        
+        init(_ index: Int) {
+            self.offset = index
+        }
+        
+        static func == (lhs: Self, rhs: Self) -> Bool {
+            lhs.offset == rhs.offset
+        }
+        
+        static func < (lhs: Self, rhs: Self) -> Bool {
+            lhs.offset < rhs.offset
+        }
+
+        static func <= (lhs: Self, rhs: Self) -> Bool {
+            lhs.offset <= rhs.offset
+        }
+
+        static func >= (lhs: Self, rhs: Self) -> Bool {
+            lhs.offset >= rhs.offset
         }
+
+        static func > (lhs: Self, rhs: Self) -> Bool {
+            lhs.offset > rhs.offset
+        }
+    }
+    
+    typealias Index = DataRow.ColumnIndex
+    
+    var startIndex: ColumnIndex {
+        ColumnIndex(self.bytes.readerIndex)
+    }
+    
+    var endIndex: ColumnIndex {
+        ColumnIndex(self.bytes.readerIndex + self.bytes.readableBytes)
+    }
+    
+    var count: Int {
+        Int(self.columnCount)
+    }
+    
+    func index(after index: ColumnIndex) -> ColumnIndex {
+        guard index < self.endIndex else {
+            preconditionFailure("index out of bounds")
+        }
+        var elementLength = Int(self.bytes.getInteger(at: index.offset, as: Int32.self)!)
+        if elementLength < 0 {
+            elementLength = 0
+        }
+        return ColumnIndex(index.offset + MemoryLayout<Int32>.size + elementLength)
```

Comment:
I don't know if the typical access pattern for a `DataRow` supports bothering with this, but wouldn't pre-calculating the buffer offsets in `.init()` then make it trivial to add `BidirectionalCollection` and `RandomAccessCollection` conformance? Or does caching the offsets itself carry enough overhead to negate the few allocations win?

Reply:
We talked about this offline:

1. Currently we expose the user `PostgresRow`. `PostgresRow` translates the DataRow into an array of `PostgresData` in its initializer. For this reason `PostgresRow` is still a `RandomAccessCollection`, without any overhead.

2. In the future we probably want to support API's like proposed in #188:

    ```swift
    let stream = try await connection.query("SELECT * FROM large_test LIMIT 2000000", logger: logger)
    for try await (column1, column2, column3) in stream.decode(Int.self, Double.self, Double.self) {
        // do something with the row
    }
    ```

    In this scenario PSQLRow doesn't have to be a `RandomAccessCollection` and we shouldn't need to pay the price for it (another allocation).